### ISSUE TITLE
The eager load associations need check association scope 

### ISSUE THREAD

**Issue by OuYangJinTing** [rails/rails#47175]
### Steps to reproduce

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org/"
  # source "https://gems.ruby-china.com/"

  gem "minitest", "~> 5.15"
  gem "minitest-reporters"
  gem "activerecord", "~> 7.0.2"
  gem "sqlite3"
end

require "minitest/autorun"
require "minitest/reporters"
require "active_record"

Minitest::Reporters.use!

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
    t.string :name, null: false
  end

  create_table :posts, force: true do |t|
    t.belongs_to :user
    t.string :title, null: false
  end
end

class User < ActiveRecord::Base
  has_many :posts, dependent: :destroy

  has_one :latest_post, -> { order(id: :desc) }, class_name: "Post"
end

class Post < ActiveRecord::Base
  belongs_to :user
end

class BugTest < Minitest::Test
  def setup
    3.times do |i|
      User.create(name: "foo-#{i + 1}").tap do |user|
        10.times do |j|
          user.posts.create(title: "bar-#{j + 1}")
        end
      end
    end
  end


  def test_correct_query
    assert_equal 10, User.first.latest_post.id
    assert_equal 10, User.includes(:latest_post).first.latest_post.id
  end

  def test_wrong_query
    assert_equal 10, User.eager_load(:latest_post).first.latest_post.id
  end
end

```

### Expected behavior

The `ActiveRecord` should raise error when try to execute `User.eager_load(:latest_post)`.
~~For example, like the `#check_eager_loadable!` method handles~~

### Actual behavior

The result of eager load is wrong

### System configuration
**Rails version**: 7

**Ruby version**: 3

**Comment by bradhuskins** [comment]
#check_eager_loadable! appears to be checking the method's arity to ensure it's a class method. Just verifying whether that's the suggested solution here.

**Comment by OuYangJinTing** [comment]
@bradhuskins Yes, as you said. `#check_eager_loadable!` can't deal with this problem. Here need to add a new method.

**Comment by bradhuskins** [comment]
Bump

**Comment by OuYangJinTing** [comment]
Bump

**Comment by skipkayhil** [comment]
```
class User < ActiveRecord::Base
  has_many :posts, dependent: :destroy

  has_one :latest_post, -> { order(id: :desc) }, class_name: "Post"
end
```

I'm curious if this usage of `has_one` is something that should be supported. Did this work with previous versions of Rails?

My expectation would be to do something like:

```
class User < ActiveRecord::Base
  has_many :posts, dependent: :destroy

  def latest_post
    posts.last # or some other scope on `posts`
  end
end
```

**Comment by OuYangJinTing** [comment]
> ```
> class User < ActiveRecord::Base
>   has_many :posts, dependent: :destroy
> 
>   has_one :latest_post, -> { order(id: :desc) }, class_name: "Post"
> end
> ```
> 
> I'm curious if this usage of `has_one` is something that should be supported. Did this work with previous versions of Rails?
> 
> My expectation would be to do something like:
> 
> ```
> class User < ActiveRecord::Base
>   has_many :posts, dependent: :destroy
> 
>   def latest_post
>     posts.last # or some other scope on `posts`
>   end
> end
> ```

To the best of my knowledge, in Rails versions 5 through 7, this feature doesn't exist.

The examples you've provided are perfectly fine, and adopting your approach can indeed resolve the issue.

However, I believe it would be even more user-friendly if this detection were supported. This could help newcomers identify issues in advance, which I find beneficial.

**Comment by OuYangJinTing** [comment]
> ```
> class User < ActiveRecord::Base
>   has_many :posts, dependent: :destroy
> 
>   has_one :latest_post, -> { order(id: :desc) }, class_name: "Post"
> end
> ```
> 
> I'm curious if this usage of `has_one` is something that should be supported. Did this work with previous versions of Rails?
> 
> My expectation would be to do something like:
> 
> ```
> class User < ActiveRecord::Base
>   has_many :posts, dependent: :destroy
> 
>   def latest_post
>     posts.last # or some other scope on `posts`
>   end
> end
> ```

@skipkayhil Thank your reply.

To the best of my knowledge, in Rails versions 5 through 7, this feature doesn't exist.

The examples you've provided are perfectly fine, and adopting your approach can indeed resolve the issue.

However, I believe it would be even more user-friendly if this detection were supported. This could help newcomers identify issues in advance, which I find beneficial.
